builtin/nomad/jobspec/platform.go

```diff
@@ -46,86 +53,246 @@ func (p *Platform) GenerationFunc() interface{} {
 	if err != nil {
-		return nil, err
+		return err
 	}
-	s.Done()
+
+	// Store our state so we can destroy it properly
+	state.Name = result.Name
+	st.Step(terminal.StatusOK, "Job registration successful")
 
 	// Wait on the allocation
-	st := ui.Status()
-	defer st.Close()
 	evalID := regResult.EvalID
 	st.Update(fmt.Sprintf("Monitoring evaluation %q", evalID))
-	if err := nomad.NewMonitor(st, client).Monitor(evalID); err != nil {
-		return nil, err
+	if err := nomad.NewMonitor(st, client.NomadClient).Monitor(evalID); err != nil {
+		return err
 	}
 	st.Step(terminal.StatusOK, "Deployment successfully rolled out!")
 
-	return &result, nil
+	return nil
 }
 
-// Destroy deletes the Nomad job.
-func (p *Platform) Destroy(
+func (p *Platform) resourceJobDestroy(
+	ctx context.Context,
+	state *Resource_Job,
+	sg terminal.StepGroup,
+	client *nomadClient,
+) error {
+	step := sg.Add("")
+	defer func() { step.Abort() }()
+	step.Update("Deleting job: %s", state.Name)
+	step.Done()
+	_, _, err := client.NomadClient.Jobs().Deregister(state.Name, true, nil)
+	return err
+}
+
+func (p *Platform) resourceJobStatus(
 	ctx context.Context,
 	log hclog.Logger,
-	deployment *nomad.Deployment,
+	sg terminal.StepGroup,
+	state *Resource_Job,
+	client *nomadClient,
+	sr *resource.StatusResponse,
 	ui terminal.UI,
 ) error {
-	// We'll update the user in real time
+	jobclient := client.NomadClient.Jobs()
+
+	s := sg.Add("Gathering health report for Nomad job...")
+	defer func() { s.Abort() }()
+
+	jobResource := sdk.StatusReport_Resource{
+		Type:                "job",
+		CategoryDisplayHint: sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER,
+	}
+	sr.Resources = append(sr.Resources, &jobResource)
+	job, _, err := jobclient.Info(state.Name, &api.QueryOptions{})
+	if err != nil {
+		return err
+	}
+
+	jobResource.Id = *job.ID
+	jobResource.Name = *job.Name
+	jobResource.CreatedTime = timestamppb.New(time.Unix(0, *job.SubmitTime))
+	stateJson, err := json.Marshal(map[string]interface{}{
+		"deployment": job,
+	})
+	jobResource.StateJson = string(stateJson)
+
+	if *job.Status == "running" {
+		jobResource.Health = sdk.StatusReport_READY
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
+	} else if *job.Status == "queued" || *job.Status == "started" {
+		jobResource.Health = sdk.StatusReport_ALIVE
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting alive!", state.Name)
+	} else if *job.Status == "completed" {
+		jobResource.Health = sdk.StatusReport_PARTIAL
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting partially available!", state.Name)
+	} else if *job.Status == "failed" || *job.Status == "lost" {
+		jobResource.Health = sdk.StatusReport_DOWN
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting down!", state.Name)
+	} else {
+		jobResource.Health = sdk.StatusReport_UNKNOWN
+		jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting unknown!", state.Name)
+	}
```

Comment:
I think we might be co-mingling several different status values here. The status of a job is only one of [running, dead, or pending](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L4021-L4025) (see also the [API docs](https://www.nomadproject.io/api-docs/jobs#status)). 

Whereas:
* An allocation's client status is one of [pending, running, complete, failed, or lost](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L9322-L9328). 
* A deployment can be [running, paused, failed, successful, cancelled, pending, blocked, or unblocking](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L8911-L8919)
* An eval can be [blocked, pending, complete, failed, canceled](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/structs/structs.go#L10370-L10376) (yes, those two objects spell canceled differently 😊 ). 
* I wasn't able to figure out which object has `queued` or `started`

The reason this granularity matters to a Waypoint user is that a Nomad job can be `running` but have a bunch of its allocations in various pending or failing states! So I suspect we want to aggregate information about the job, deployment, and its allocations and then present that as a top-level Waypoint job status.

Reply:
@tgross thank you for advising on this! It looks like "queued" and "started" may have initially been sourced from the [TaskGroupSummary struct](https://github.com/hashicorp/nomad/blob/35c22bcb6c31d3605d9fe9f2a02cefee04e6c4b1/api/jobs.go#L1003-L1010), but I agree that it would make sense to do some kind of top-level status. Here's what I'm thinking...

If a deployment's status is "completed", skip the eval check and do alloc health check.

If a deployment's status is running, check eval ID returned. Since the eval may return a "blocked", "failed" or "canceled" status, that would allow us to skip the alloc evaluation and jump straight to a final unhealthy status.

If a deployment's status is "paused", "failed", or "blocked", early exit (skip alloc health check) with a final status.

If a deployment's status is "pending" or "unblocking", I'm a bit unsure how to proceed here... I'm thinking either provide an UNKNOWN status, or wait for some period of time and check again for a different status until we hit a timeout on the status check. 

For the alloc health check, if all allocs are "running", we go with "READY". If some are "running" and some are "pending" or "failed", we report "PARTIAL". If all are "lost" or "failed", we report "DOWN". I feel like if we have allocs in the "complete" state, then technically the job is "DOWN", but that could possibly be perceived as unhealthy to Waypoint users, even if the job ran to completion as intended (like a batch job).

Final status options for Waypoint, for reference:
- UNKNOWN: We cannot determine the status of the resource
- ALIVE: The resource exists but might not be ready to handle requests
- READY: The resource exists and is ready to handle requests
- DOWN: The resource is down and not responding
- MISSING: We're expecting it to exist, but it does not
- PARTIAL: Some resources in deployment are OK, others are not OK

Reply:
> If a deployment's status is "completed", skip the eval check and do alloc health check.

A deployment could be "complete" with failed-but-replaced allocations. Or you could complete the deploy and then immediately have an unrelated alloc failure. I'm not sure of the semantics the Waypoint user would want there in terms of the output of this command?

> If a deployment's status is running, check eval ID returned. Since the eval may return a "blocked", "failed" or "canceled" status, that would allow us to skip the alloc evaluation and jump straight to a final unhealthy status.

If the deployment is "running", the eval that led to that deployment can only be "complete". The scheduler may have _also_ spawned a new blocked eval if the plan was only partially applied. Ex. if you submit a job that needs 20 allocs but can only fit 18, we start the deployment for those 18 and then create a new "blocked" eval to re-plan the remaining 2 allocations once more resources become available. In that case it sounds like that'd be `PARTIAL`?

> If a deployment's status is "paused", "failed", or "blocked", early exit (skip alloc health check) with a final status.
>
> If a deployment's status is "pending" or "unblocking", I'm a bit unsure how to proceed here... I'm thinking either provide an UNKNOWN status, or wait for some period of time and check again for a different status until we hit a timeout on the status check.

A failed deployment is definitely a failure (`DOWN`?). 

A paused deployment has been manually paused by the operator with [`nomad deployment pause`](https://www.nomadproject.io/docs/commands/deployment/pause). Not sure what's the best Waypoint state to reflect that.

The "blocked", "pending", "unblocking" states are kind of exotic. Those are for [multi-region deployments](https://learn.hashicorp.com/tutorials/nomad/multiregion-deployments). I'm not sure if Waypoint supports federated Nomad regions at all yet, but if you got those you'd want to poll on them until they become terminal.

> For the alloc health check, if all allocs are "running", we go with "READY". If some are "running" and some are "pending" or "failed", we report "PARTIAL". If all are "lost" or "failed", we report "DOWN". I feel like if we have allocs in the "complete" state, then technically the job is "DOWN", but that could possibly be perceived as unhealthy to Waypoint users, even if the job ran to completion as intended (like a batch job).

That all seems reasonable to me.

Speaking of non-service jobs, you might also want to special case system jobs because they don't use deployments.

Reply:
@tgross after further evaluation, I took a different direction here. It occurred to me that since Waypoint continuously polls for app status via the `resourceJobStatus` function (see: #1612), it may not be best for us to rely on the deployment or evaluation from the deployment event, since the information they provide will become stale after the deployment event. Because of that, in the update I made, we now check for the 3 job statuses, running/pending/dead, and only if the job is running do we dig deeper into the status of the job's allocs:

```go
		if running == currentJobVersionAllocs {
			jobResource.Health = sdk.StatusReport_READY
			jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
		} else if running > 0 && (complete > 0 || failed > 0 || pending > 0 || lost > 0) {
			jobResource.Health = sdk.StatusReport_PARTIAL
			jobResource.HealthMessage = fmt.Sprintf("Some allocations are running for Job %q!", state.Name)
		} else if (complete + pending + failed + lost) == currentJobVersionAllocs {
			jobResource.Health = sdk.StatusReport_DOWN
			jobResource.HealthMessage = fmt.Sprintf("No allocations for Job %q are running!", state.Name)
		}
```
https://github.com/hashicorp/waypoint/pull/2938/commits/608e787bb24121939dc603c72261ba7a6c0d15d8#diff-15311425528bec1ed88b93dfce5b4fa4ef9e23779e2d1f2229a55e976ccbb913R203-R230

Reply:
@tgross I've updated the check the list of evals for an alloc, and if there's any blocked evals, but all allocs are running then the status becomes PARTIAL.

Eval check:
```go
	if *job.Status == "running" {
		// Get list of evaluations for job
		evals, _, err := jobClient.Evaluations(*job.ID, q)
		evalStatus := "unblocked"
		for _, eval := range evals {
			switch eval.Status {
			case structs.EvalStatusBlocked:
				evalStatus = "blocked"
				break
			case structs.EvalStatusPending:
				break
			case structs.EvalStatusComplete:
				break
			case structs.EvalStatusFailed:
				break
			case structs.EvalStatusCancelled:
				break
			default:
				evalStatus = "unknown"
			}
		}
```

Alloc check & eval check:
```go
		if running == currentJobVersionAllocs && evalStatus == "unblocked" {
			jobResource.Health = sdk.StatusReport_READY
			jobResource.HealthMessage = fmt.Sprintf("Job %q is reporting ready!", state.Name)
		} else if running == currentJobVersionAllocs && evalStatus == "blocked" {
			jobResource.Health = sdk.StatusReport_PARTIAL
			jobResource.HealthMessage = fmt.Sprintf("Allocs for job %q are running, but there is at least one blocked evaluation!", state.Name)
		}
```